### ISSUE TITLE
Proposal: Delegates for getters and setters

### ISSUE THREAD

**Issue by jamesharling** [dotnet/csharplang#2467]
Currently, if a property's getter or setter features logic then a backing field must be created manually, and the getter/setter should reference that field. For example, consider the following where we only wish to set the value if it is not null, empty or white space:

```
private string name;

public string Name
{
    get => this.name;
    set => this.name = String.IsNullOrWhiteSpace(value) ? null : value;
}
```

It is proposed to add syntactic sugar to negate the explicit definition of the backing field. This sugar would take the form of a delegate, with the following signatures:

* getter
  * `delegate TProperty Get(TProperty current)`
* setter
  * `delegate TProperty Set(TProperty new)` or
  * `delegate TProperty Set(TProperty current, TProperty new)`

A corresponding backing field of type `TProperty` would be created by the compiler.

It would then be possible to combine expression-bodied properties with these delegates, simplifying our initial example:

```
public string Name
{
    get => (current) => current;
    set => (new) => String.IsNullOrWhiteSpace(new) ? null : new;
}
```

If more complex logic was required, a statement body could be used.

An implicit limitation is that the automatically-generated backing field would be of the same type as the defined property - this restriction could be negated by using explicit backing fields.

**Comment by miloush** [comment]
```C#
public Func<string, string> Name
{
    get => (current) => current;
}
```
already compiles and would conflict with the proposed syntax.

**Comment by HaloFour** [comment]
Check out #140

**Comment by jamesharling** [comment]
@miloush the proposed could be limited to types of `TProperty` that are not themselves a delegate

@HaloFour I had seen this, and wondered if similar functionality could be achieved without the addition of a new keyword

**Comment by svick** [comment]
@jamesharling I find something similar to https://github.com/dotnet/csharplang/issues/140 much easier to understand than this proposal, because:

1. It's very close to syntax that already means something.
2. It's not consistent with how setters currently work, where the new value is always accessed as `value`.

**Comment by jamesharling** [comment]
Closed, in favour of #140
